File: python/taquin.py
#!/usr/local/bin/python3
# -*-coding:utf-8 -*

from random import shuffle
from math import ceil
from collections import OrderedDict
import time

Infinity = 10000000000

class Taquin:
	def __init__(self, environment, previous=None, move=None):
		self.environment = environment
		self.previous = previous
		self.inv = None
		self.dis = None
		self.man = None
		self.h = None
		if previous == None:
			self.path = "_"
			self.g = 0
			self.sequence = self.magic(1)
		else:
			self.path = previous.path + move
			self.g = previous.g + 1
			self.sequence = previous.sequence.copy()
			self.moveTile(move)
			self.inv,self.dis,self.man,self.h = self.details()
		self.moves = self.findMoves()
		self.f = self.h + self.g
	def coordinates(self, content=0):
		width = self.environment.sizes[0]
		if isinstance(content, list): return (width * content[1]) + content[0]
		else:
			index = self.sequence.index(content)
			y = ceil((index + 1) / width) - 1
			x = index - (y * width)
			return [x, y]

	def details(self):
		width, length = self.environment.sizes
		weightings = self.environment.weightings
		sequence = self.sequence
		inv = 0
		dis = 0
		man = 0
		h = 0
		for weighting in weightings:
			k = 0
			stepH = 0
			for i in range(0,length):
				stepMan = 0
				if weighting == weightings[0]:
					for j in range(i+1,length):
						if sequence[i] != 0 and sequence[j] != 0 and sequence[i] > sequence[j]: inv += 1
					if sequence[i] != 0 and sequence[i] != (i+1):
						dis += 1
				if i > 0:
					pos = self.coordinates(i)
					x = i % width
					coords = (((width - 1) if x == 0 else (x - 1)), ceil(i / width) - 1)
					stepMan += (abs(pos[0] - coords[0]) + abs(pos[1] - coords[1]))
					if weighting == weightings[0]: man += stepMan
					stepH += weighting[0][k] * stepMan
					k += 1
			stepH /= weighting[1]
			if weighting[2] == 7: h += dis
			else: h += stepH
		return [inv,dis,man,h]

	def findMoves(self,flex=False):
		limit = self.environment.sizes[0] - 1
		coords = self.coordinates()
		last = self.path[self.g]
		moves = []
		if coords[0] != 0 	  and (last != 'L' or flex): moves.append('R')
		if coords[0] != limit and (last != 'R' or flex): moves.append('L')
		if coords[1] != 0	  and (last != 'U' or flex): moves.append('D')
		if coords[1] != limit and (last != 'D' or flex): moves.append('U')
		return moves
	def moveTile(self, move):
		sequence = self.sequence
		width = self.environment.sizes[0]
		x = self.coordinates(self.coordinates())
		if move == 'R': y = x - 1
		if move == 'L': y = x + 1
		if move == 'D': y = x - width
		if move == 'U': y = x + width
		sequence[x] = sequence[y]
		sequence[y] = 0
	def valid(self):
		width = self.environment.sizes[0]
		self.inv,self.dis,self.man,self.h = self.details()
		row = abs(self.coordinates()[1] - width)
		return True if (((width % 2 == 1) and (self.inv % 2 == 0)) or ((width % 2 == 0) and ((row % 2 == 1) == (self.inv % 2 == 0)))) else False
	def children(self):
		childList = []
		for move in self.moves:
			child = Taquin(self.environment,self,move)
			if child.h == 0: return child
			i = 0
			while (i < len(childList)):
				if (child.f < childList[i].f): break
				i += 1
			childList.insert(i,child)
		return childList
	def magic(self, rand=0):
		length = self.environment.sizes[1]
		sequence = [0]*length
		for i in range(1, length): sequence[i-1] = i
		if rand == 1:
			shuffle(sequence)
			self.sequence = sequence
			while not self.valid():
				shuffle(sequence)
				self.sequence = sequence
		return sequence
	def traceroute(self):
		path = [self]
		while (isinstance(path[0].previous,Taquin)): path.insert(0,path[0].previous)
		return path
	def __repr__(self):
		printable = ""
		printable += "\n"
		printable += "Taquin :\n"
		printable += ("|  seq. .. : {}\n").format(self.sequence)
		printable += ("|  path .. : {}\n").format(self.path)
		printable += ("|  inv. .. : {}\n").format(self.inv)
		printable += ("|  man. .. : {}\n").format(self.man)
		printable += ("|  moves . : {}\n").format(self.moves)
		printable += ("|  g ..... : {}\n").format(self.g)
		printable += ("|  h ..... : {}\n").format(self.h)
		printable += ("|  f ..... : {}\n").format(self.f)
		return printable


class Environment:
	def __init__(self,width,choices=None):
		self.sizes = (width,width*width)
		self.choices = choices
		self.weightings = self.getWeightings(choices)
		self.moves = [Taquin(self)]
		self.end = []
	def getWeightings(self,choices):
		width = self.sizes[0]
		length = self.sizes[1] - 1
		if (choices == None): choices = [5]
		weightings = []
		weight = length
		for index in choices:
			rho = (4 if index % 2 != 0 else 1)
			pi = [0] * length
			if index == 1:
				if width == 3: pi = [36, 12, 12, 4, 1, 1, 4, 1]
			if index == 2 or index == 3:
				pi = [(length+1) - i for i in range(1,length+1)]
			if index == 4 or index == 5:
				pi = [0] * length
				weight = length
				for i in range(width-1):
					j = 0
					while pi[j] != 0:
						j += 1
					k = 0
					while k < width-i:
						pi[j] = weight
						j += 1
						weight -= 1
						k += 1
					j += i
					pi[j] = weight
					weight -= 1
					j += width
					while j < length - 1 :
						pi[j] = weight
						weight -= 1
						j += width
			if index == 6:
				pi = [1] * length
			if not index in [1,2,3,4,5,6]:
				pass
			weightings.append((pi,rho,index))
		return weightings
	def correct(self):
		for move in self.moves:
			move.inv,move.dis,move.man,move.h = move.details()
			move.f = move.g + move.h
			move.moves = move.findMoves( True )
	
	@staticmethod
	def inArray(taquin,array):
		for element in array:
			if element.sequence == taquin.sequence: return True
		return False
	
	
	def aStar(self):
		print(self.moves[-1])
		queue = OrderedDict()
		queue[self.moves[-1].f] = [self.moves[-1]]
		while (True):
			k = list(queue.keys())[0]
			shouldBeExpanded = queue[k][0]
			del queue[k][0]
			if queue[k] == []: del queue[k]
			children = shouldBeExpanded.children()
			if isinstance(children,Taquin):
				print(children)
				self.end.append(children)
				return self.end[-1]
			else:
				for child in children :
					if(child.f in queue): queue[child.f].append(child)
					else: queue[child.f] = [child]
				queue = OrderedDict( sorted( queue.items(), key=lambda t: t[0]))
	
	def charlotte(self):
		root = self.moves[-1]
		bound = root.h
		path = [root]

		def search(path,g,bound):
			node = path[-1]
			f = g + node.h
			if f > bound: return f
			minimum = Infinity
			children = node.children()
			if isinstance(children,Taquin):
				path.append(children)
				return children
			else:
				for child in children:
					if not child.environment.inArray(child,path):
						path.append(child)
						t = search(path,g+1,bound)
						if isinstance(t,Taquin): return t
						if t < minimum: minimum = t
						path.pop()
			return minimum

		while (True):
			t = search(path,0,bound)
			if isinstance(t,Taquin):
				print(t)
				self.end.append(t)
				return t
			if t == Infinity: return False
			bound = t

	
	def expand(self,function,decomposition=0):
		if (decomposition==0):
			print("\n\n")
			start = time.time()
			print(("Heuristiques utilisées : {}").format(self.choices))
			result = function()
			print(("Duration : {}").format(time.time() - start))
			print("\n\n")
			return result
		else:
			results = []
			decomposition = self.weightings.copy()
			for weighting in decomposition:
				print("\n")
				start = time.time()
				print(("Heuristiques utilisées : {}").format(weighting[2]))
				self.weightings = [weighting]
				self.correct()
				results.append(function())
				print(("Duration : {}").format(time.time() - start))
				print("\n\n.........................................\n")
			return results
	
	
	
	def play(self,move):
		self.moves.append(Taquin(self,self.moves[-1],move))
		return self.moves[-1]





#class __main__:
#	width = int(input(">>> Taille du taquin ?\n>>> "))
#	choices = str(input(">>> Heuristiques ?\n>>> Entrez les numéros séparés par des espaces.\n>>> "))
#	decomposition = 0
#	if len(choices) == 1: choices = [int(choices)]
#	else:
#		choices = choices.split(' ')
#		for index,choice in enumerate(choices): choices[index] = int(choice)
#		decomposition = int(input(">>> Voulez-vous associer les heuristiques [0] ou dissocier les exécutions [1] ?\n>>> "))
#	a = Environment(width,choices)
#	while(a.moves[-1].h != 0):
#		print(a.moves[-1])
#		move = "_"
#		while not move in ["R","L","D","U","E"]:
#			move = str(input((">>> Dans quel direction voulez vous aller ? {}\n>>> Ou alors peut-être voulez-vous explorer ? ['E']\n>>> ").format(a.moves[-1].moves)))
#		if move in a.moves[-1].moves:
#			a.play(move)
#		elif move == "E":
#			a.expand(a.charlotte,decomposition)
#			exit(0)

File: docs/taquin.js
class Taquin {
	constructor( environment, previous = undefined, move = undefined ) {
		this.environment = environment;
		this.previous = previous;
		this.inv = undefined;
		this.dis = undefined;
		this.man = undefined;
		this.h = undefined;
		if ( previous == undefined ) {
			this.path = "";
			this.g = 0;
			this.sequence = this.magic( 1 );
		} else {
			this.path = previous.path + move;
			this.g = this.previous.g + 1;
			this.sequence = previous.sequence.slice();
			this.moveTile( move );
			[this.inv,this.dis,this.man,this.h] = this.details();
		}
		this.moves = this.findMoves();
		this.f = this.h + this.g;
	}
	coordinates( content = 0 ) {
		let width = this.environment.sizes[0];
		if ( content instanceof Object ) {
			return ( width * content[1] ) + content[0];
		} else {
			let index = this.sequence.indexOf( content );
			let y = Math.ceil( ( index + 1 ) / width ) - 1;
			let x = index - ( y * width );
			return [x, y];
		}
	}
	details() {
		let [width,length] = this.environment.sizes,
		weightings = this.environment.weightings,
		seq = this.sequence,
		inv = 0,
		dis = 0,
		man = 0,
		h = 0;
		for ( let weighting of weightings ) {
			let k = 0,
			stepH = 0;
			for ( let i = 0 ; i < length ; i++ ) {
				let stepMan = 0;
				if ( weighting == weightings[0] ) {
					for ( let j = i+1 ; j < length ; j++ ) {
						if ( seq[i] != 0 && seq[j] != 0 && seq[i] > seq[j] ) {
							inv++;
						}
					}
					if ( seq[i] != 0 && seq[i] != ( i + 1 ) ) {
						dis++;
					}
				}
				if ( i > 0 ) {
					let pos = this.coordinates( i ),
					x = i % width,
					coords = [( ( x == 0 ) ? ( width - 1 ) : ( x - 1 ) ), ( Math.ceil( i / width ) - 1 )];
					stepMan += ( Math.abs( pos[0] - coords[0] ) + Math.abs( pos[1] - coords[1] ) );
					if ( weighting == weightings[0] ) {
						man += stepMan;
					}
					stepH += weighting[0][k] * stepMan;
					k++;
				}
			}
			if ( weighting[2] == 7 ) {
				stepH += dis;
			}
			stepH = parseInt(stepH / weighting[1]);
			h += stepH;
		}
		h = parseInt( h / weightings.length );
		return [inv,dis,man,h];
	}
	findMoves( flex = false ) {
		let limit = this.environment.sizes[0] - 1;
		let coords = this.coordinates();
		let last = this.path[this.g];
		let moves = [];
		if ( coords[0] != 0 && ( last != 'L' || flex ) ) { moves.push( 'R' ); }
		if ( coords[0] != limit && ( last != 'R' || flex ) ) { moves.push( 'L' ); }
		if ( coords[1] != 0 && ( last != 'U' || flex ) ) { moves.push( 'D' ); }
		if ( coords[1] != limit && ( last != 'D' || flex ) ) { moves.push( 'U' ); }
		return moves;
	}
	moveTile( move ) {
		let seq = this.sequence,
		width = this.environment.sizes[0],
		x = this.coordinates( this.coordinates() ),
		y;
		if ( move == 'R' ) { y = x - 1; }
		if ( move == 'L' ) { y = x + 1; }
		if ( move == 'D' ) { y = x - width; }
		if ( move == 'U' ) { y = x + width; }
		seq[x] = seq[y];
		seq[y] = 0;
	}
	valid() {
		let width = this.environment.sizes[0];
		[this.inv,this.dis,this.man,this.h] = this.details();
		let row = Math.abs( this.coordinates()[1] - width );
		return ( ( width % 2 == 1 ) && ( this.inv % 2 == 0 ) ) || ( ( width % 2 == 0 ) && ( ( row % 2 == 1 ) == ( this.inv % 2 == 0 ) ) ) ? true : false;
	}
	children() {
		let childList = [];
		for ( let move of this.moves ) {
			let child = new Taquin( this.environment, this, move );
			if ( child.dis == 0 ) {
				return child;
			}
			let i = 0;
			for (i; i < childList.length; i++) {
				if (child.f < childList[i].f) {
					break;
				}
			}
			childList.splice(i,0,child);
		}
		return childList;
	}
	magic( rand = 0 ) {
		let length = this.environment.sizes[1],
		seq = new Array( length ).fill(0);
		for ( let i = 1 ; i < length ; i++ ) {
			seq[i-1] = i;
		}
		if ( rand == 1 ) {
			do {
				seq.shuffle();
				this.sequence = seq;
			} while ( !this.valid() );
		}
		return seq;
	}
	traceroute() {
		let path = [this];
		while ( path[0].previous instanceof Taquin ) {
			path.unshift( path[0].previous );
		}
		return path;
	}
}

class Environment {
	constructor( width, choices = undefined ) {
		this.createdTaquins = 0;
		this._sizes = [width, width * width];
		this.choices = choices;
		this._weightings = this.getWeightings( choices );
		this.moves = [new Taquin( this )];
		this.end = [];
	}
	get sizes() { return this._sizes; }
	getWeightings( choices ) {
		let width = this.sizes[0],
		length = this.sizes[1] - 1;
		if ( choices == undefined ) {
			choices = [5];
		}
		let weightings = [],
		weight = length;
		for ( let index of choices ) {
			let rho = index % 2 != 0 ? 4 : 1;
			let pi = new Array(length).fill(0);
			switch (index) {
				case 1:
					if ( width == 3 ) {
						pi = [36, 12, 12, 4, 1, 1, 4, 1];
					} else {
						for ( let y = 0 ; y < width ; y++ ) {
							for ( let x = 0 ; x < width ; x++ ) {
								if ( x == y == width-1 ) {
									continue;
								} else {
									if ( x == y == 0 ) {
										pi[0] = width * ( width * 3 );
										x++;
									}
									if ( y == 0 ) {
										while ( x < width ) {
											pi[x++] = width * 3;
										}
									} else {
										if ( x == 0 ) {
											pi[y*width] = width * 2;
										} else {
											pi[y*width+x] = width - y;
										}
									}
								}
							}
						}
					}
					break;
				case 2:
				case 3:
					pi = Array.from( Array( length ), (_e,i) => ( length+1 ) - ( i+1 ) );
					break;
				case 4:
				case 5:
					for (let i = 0; i < ( width - 1 ) ; i++ ) {
						let j = 0;
						while ( pi[j] != 0 ) {
							j++;
						}
						for ( let k = 0 ; k < ( width - i ) ; k++ ) {
							pi[j++] = weight--;
						}
						j += i;
						pi[j] = weight--;
						j += width;
						while ( j < length - 1 ) {
							pi[j] = weight--;
							j += width;
						}
					}
					break;
				case 6:
					pi = Array( length ).fill( 1 );
					rho = 1 / ((width - 3) + 1);
					break;
				case 7:
					break;
				case 8:
					let mid = Math.floor(length/2);
					for (let i = 0; i < mid ; i++) {
						pi[i] = mid - i;
					}
					if (length % 2 == 1) {
						pi[mid] = length;
						mid++;
					}
					for (let i = mid; i < length; i++) {
						pi[i] = i+1;
					}
					rho = 2.5;
					break;
				case 9:
					rho = 2;
					let j = 1;
					for (let i = 0; i < length; i++) {
						pi[i] = Math.abs(Math.floor(length/2) - (Math.floor((j-1)/2)));
						if (i < length-1) {
							i++;
							pi[i] = Math.abs(Math.floor(length/2) - (Math.floor((j-1)/2)));
						}
						j++;
					}
					if (length % 2 == 1) {
						pi[length-1] = 1;
					}
					pi.shuffle();
					break;
				default:
					break;
			}
			weightings.push( [pi,rho,index] );
		}
		return weightings;
	}
	get weightings() { return this._weightings; }
	set weightings( choices ) { this._weightings = this.getWeightings( choices ); }
	correct() {
		for ( let move of this.moves ) {
			[move.inv,move.dis,move.man,move.h] = move.details();
			move.f = move.g + move.h;
			move.moves = move.findMoves( true );
		}
	}
	inArray(taquin,array) {
		for (let element of array) {
			if (element.sequence == taquin.sequence) {
				return true;
			}
		}
		return false;
	}
	aStar() {

		let explored = new Map();
		let queue = new Map();
		queue.set(this.moves.last().f,[this.moves.last()]);

		while (true) {
			let k = Array.from( queue.keys() )[0];
			let kArray = queue.get( k );
			let shouldBeExpanded = kArray.shift();

			explored.set(shouldBeExpanded.sequence.toString(),shouldBeExpanded);

			if ( kArray.length == 0 ) {
				queue.delete(k);
			} else {
				queue.set( k, kArray );
			}

			let children = shouldBeExpanded.children();
			if (children instanceof Taquin) {
				this.end.push(children);
				return this.end.last();
			} else {
				for (let child of children) {
					let sequenceString = child.sequence.toString();
					if (explored.has(sequenceString)) {
						if (explored.get(sequenceString).f < child.f) {
							children.splice(children.indexOf(child),1);
						} else {
							explored.delete(sequenceString);
						}
					} else if (queue.has(child.f)) {
						let cArray = queue.get(child.f);
						cArray.push(child);
						queue.set(child.f,cArray);
					} else {
						queue.set(child.f,[child]);
					}
				}
			}
			let sortedArray = Array.from(queue.keys());
			sortedArray.sort((a,b)=>a-b);
			let secondaryQueue = new Map();
			for (let key of sortedArray) {
				secondaryQueue.set(key,queue.get(key));
				queue.delete(key);
			}
			queue = secondaryQueue;
		}
	}

	idaStar() {
		let root = this.moves.last();
		let bound = root.h;
		let path = [root];
		function search(path,g,bound) {
			let node = path.last();
			let f = g + node.h;
			if (f > bound) { return f; }
			let minimum = Infinity;
			let children = node.children();
			if (children instanceof Taquin) {
				path.push(children);
				return children;
			} else {
				for (let child of node.children()) {
					if (!child.environment.inArray(child,path)) {
						path.push(child);
						let t = search(path,g+1,bound);
						if (t instanceof Taquin) {return t;}
						if (t < minimum) { minimum = t;}
						path.pop();
					}
				}
			}
			return minimum;
		}
		while (true) {
			let t = search(path,0,bound);
			if (t instanceof Taquin) {
				this.end.push(t);
				return t;
			}
			if (t == Infinity) {return false;}
			bound = t;
		}
	}

	hal() {
		let root = this.moves.last(),
		list = [root],
		explored = [];
		for (let n of list) {
			let children = n.children();
			if (children instanceof Taquin) {
				this.end.push(children);
				return children;
			}
			else {
				for (let i = children.length-1; i >= 0; i--) {
					if (!this.inArray(children[i],explored)) {
						list.unshift(children.pop());
					} else {
						children.pop();
					}
				}
			}
			let i = 0;
			for (i; i < explored.length; i++) {
				if (n.f < explored[i].f) {
					break;
				}
			}
			explored.splice(i,0,n);
			list.splice(list.indexOf(n),1);
		}
		return false;
	}

	expand( func ) {
		this.correct();
		return this[func]();
	}
	play(move) {
		lastTaquin = this.moves.last(),
		newTaquin = new Taquin(this,lastTaquin,move);
		newTaquin.moves = newTaquin.findMoves(true);
		this.moves.push(newTaquin);
		display.taquin.play();
		if (this.moves.last().dis == 0) {
			document.body.classList.toggle("win");
		}
	}
}
